backend/hct_mis_api/one_time_scripts/program_cycle_data_migration.py

```diff
@@ -0,0 +1,173 @@
+            if cycle.end_date < cycle.start_date:
+                cycle.end_date = cycle.start_date
+            try:
+                cycle.save()
+            except ValidationError:
+                # if validation error just save one day cycle
+                cycle.end_date = cycle.start_date
+                cycle.save()
+
+            previous_cycle = cycle
+
+
+def generate_unique_cycle_title(start_date: str) -> str:
+    # add to the cycle title just random 4 digits
+    while True:
+        cycle_name = f"Cycle {start_date} ({str(randint(1111, 9999))})"
+        if not ProgramCycle.objects.filter(title=cycle_name).exists():
+            return cycle_name
+
+
+def create_new_program_cycle(
+    program_id: str, status: str, start_date: str, end_date: Optional[str] = None
+) -> ProgramCycle:
+    return ProgramCycle.objects.create(
+        title=generate_unique_cycle_title(start_date),
+        program_id=program_id,
+        status=status,
+        start_date=start_date,
+        end_date=end_date,
+        created_by=None,
+    )
+
+
+def processing_with_finished_program(payment_plan: PaymentPlan, start_date: str, end_date: str) -> None:
+    # update if exists or create new cycle
+    if ProgramCycle.objects.filter(program_id=payment_plan.program_id).exists():
+        cycle = ProgramCycle.objects.filter(program_id=payment_plan.program_id).first()
+        if cycle.start_date != start_date:
+            cycle.start_date = start_date
+        if cycle.end_date != end_date:
+            cycle.end_date = end_date
+        if cycle.status != ProgramCycle.FINISHED:
+            cycle.status = ProgramCycle.FINISHED
+        cycle.save(update_fields=["start_date", "end_date", "status"])
+    else:
+        cycle = create_new_program_cycle(str(payment_plan.program_id), ProgramCycle.FINISHED, start_date, end_date)
+
+    # update TP
+    TargetPopulation.objects.filter(id=payment_plan.target_population_id).update(program_cycle=cycle)
+    # update Payment Plan
+    PaymentPlan.objects.filter(id=payment_plan.id).update(program_cycle=cycle)
+
+
+def processing_with_active_program(payment_plan: PaymentPlan) -> None:
+    logger.info(f"** ** Processing Payment Plan {payment_plan.unicef_id}")
+    payment_plan_start_date = str(payment_plan.start_date.date())
+    payment_plan_end_date = str(payment_plan.end_date.date())
+
+    if ProgramCycle.objects.filter(program_id=payment_plan.program_id).exists():
+        cycle = ProgramCycle.objects.filter(program_id=payment_plan.program_id).first()
+    else:
+        cycle = create_new_program_cycle(
+            str(payment_plan.program_id), ProgramCycle.ACTIVE, payment_plan_start_date, payment_plan_end_date
+        )
+
+    # check if any conflicts in the cycle
+    new_hh_ids = set(payment_plan.eligible_payments.values_list("household_id", flat=True))
+    for comparing_with_payment_plan in PaymentPlan.objects.filter(program_cycle=cycle):
```

Comment:
maybe we should exclude current `payment_plan` from the query?

Reply:
going to refactor this script